deps/GraphBLAS/Source/GxB_Matrix_tuple_iter.c

```diff
@@ -13,59 +24,118 @@ static inline void _EmptyIterator
 	iter->A = A ;
 	iter->nnz_idx = 0 ;
 	iter->row_idx = 0 ;
 	iter->nrows = nrows ;
 	iter->p = A->p[0] ;
 	if(iter->sparsity_type == GxB_HYPERSPARSE) {
-		iter->sparse_row_idx = 0 ;
-		iter->h_size = A->nvec ;
+		_MatrixTupleIter_init_hypersparse_fields(iter, 0, A->nvec);
 	}
 
 	return GrB_SUCCESS ;
 }
 
-// Create a new iterator
-GrB_Info GxB_MatrixTupleIter_new
+// finds the start row index in Ah for HYPERSPARSE matrix,
+// returns true if found else false.
+static inline bool _find_minimal_row_in_Ah_greater_or_equal_to_rowIdx
 (
-	GxB_MatrixTupleIter **iter,     // list of iterators to create
-	const GrB_Matrix A              // matrix to iterate over
+	_GxB_MatrixTupleIter *iter,
+	GrB_Index rowIdx,
+	GrB_Index *result
 ) {
-	GB_WHERE(A, "GxB_MatrixTupleIter_new (iter, A)") ;
-	GB_RETURN_IF_NULL_OR_FAULTY(A) ;
-	*iter = GB_MALLOC(1, GxB_MatrixTupleIter) ;
-	(*iter)->iterators_idx = 0;
-	(*iter)->iterators_size = 1;
+	GB_WHERE1("_find_minimal_row_in_Ah_greater_or_equal_to_rowIdx (iter, rowIdx, result)") ;
+	if(!iter->h_size) return false ; // no rows in Ah return not found
+	GrB_Matrix A = iter->A ;
+	bool found ;
+	GrB_Index l = 0, r = iter->h_size - 1 ;
 
-	return GxB_MatrixTupleIter_init(&(*iter)->iterators[0], A) ;
+	GB_BINARY_SEARCH(rowIdx, A->h, l, r, found) ;
+	if(found) {
+		*result = l ;
+		return found ;
+	} else if(A->h[l] > rowIdx) {
+		// rowIdx not found, look for the minimal row index which is greater than rowIdx
+		// this can be located in h[l] or h[l+1]
+		*result = l ;
+		return true ;
+	} else if(l + 1 < iter->h_size) {
+		if(A->h[l + 1] <= rowIdx) {
+			GB_ERROR(GrB_INVALID_VALUE,
+				"row index " GBu " must be > than rowIdx" GBu,
+				A->h[l + 1], rowIdx) ;
+		}
+		*result = l+1 ;
+		return true ;
+	}
+
+	// row index matching criteria wasn't found
+	return false ;
 }
 
-// finds the row index in Ah for HYPERSPARSE matrix,
+// finds the end row index in Ah for HYPERSPARSE matrix,
 // returns true if found else false.
-static inline bool _find_row_index_in_Ah
+static inline bool _find_maximal_row_in_Ah_smaller_or_equal_to_rowIdx
 (
 	_GxB_MatrixTupleIter *iter,
 	GrB_Index rowIdx,
 	GrB_Index *result
 ) {
-	if(!result || !iter || !iter->h_size) return false ;
-	GrB_Index l = 0, h = iter->h_size - 1, m ;
+	GB_WHERE1("_find_maximal_row_in_Ah_smaller_or_equal_to_rowIdx (iter, rowIdx, result)") ;
+	if(!iter->h_size) return false ; // no rows in Ah return not found
+	GrB_Matrix A = iter->A ;
+	bool found ;
+	GrB_Index l = 0, r = iter->h_size - 1 ;
+
+	GB_BINARY_SEARCH(rowIdx, A->h, l, r, found) ;
+	if(found) {
+		*result = l ;
+		return found ;
+	} else if(A->h[l] < rowIdx) {
+		// rowIdx not found, look for the maximal row index which is smaller than rowIdx
+		// this can be located in h[l] or h[l-1]
+		*result = l ;
+		return true ;
+	} else if(l > 0) {
+		if(A->h[l - 1] >= rowIdx) {
+			GB_ERROR(GrB_INVALID_VALUE,
+				"row index " GBu " must be < than rowIdx" GBu,
+				A->h[l + 1], rowIdx) ;
+		}
```

Comment:
Feels like this should be an assertion, `ASSERT`
Please note `GB_ERROR` returns GrB_Info, this function returns `bool`.